### PR TITLE
Add a flag for find_unused_parameters

### LINKED ISSUE #9802: [trainer] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters
When running DDP:
```
rm -r /tmp/test-clm; CUDA_VISIBLE_DEVICES=0,1 python -m torch.distributed.launch --nproc_per_node 2 \
run_clm.py --model_name_or_path gpt2 --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 \
--do_train --output_dir /tmp/test-clm --per_device_train_batch_size 4 --max_steps 200
```
I get:

> [W reducer.cpp:1050] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters. This flag results in an extra traversal of the autograd graph every iteration, which can adversely affect performance. If your model indeed never has any unused parameters, consider turning this flag off. Note that this warning may be a false positive your model has flow control causing later iterations to have unused parameters. (function operator())

but it's not possible to turn it off from the trainer. i.e. it's hardwired.

@sgugger

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds a flag to control whether `find_unused_parameters` is set to `True` or not in DDP training, while keeping the current behavior as default to avoid any breaking change.
Fixes #9802

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -761,18 +761,20 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
         elif is_sagemaker_distributed_available():
             model = DDP(model, device_ids=[dist.get_local_rank()], broadcast_buffers=False)
         elif self.args.local_rank != -1:
+            if self.args.ddp_find_unused_parameters is not None:
+                find_unused_parameters = self.args.ddp_find_unused_parameters
+            elif isinstance(model, PreTrainedModel):
+                # find_unused_parameters breaks checkpointing as per
+                # https://github.com/huggingface/transformers/pull/4659#issuecomment-643356021
+                find_unused_parameters = getattr(model.config, "gradient_checkpointing", False)
```
Comment (stas00) [inline src/transformers/trainer.py:769]:
```suggestion
                find_unused_parameters = not getattr(model.config, "gradient_checkpointing", False)
```

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -240,6 +240,10 @@ class TrainingArguments:
         report_to (:obj:`List[str]`, `optional`, defaults to the list of integrations platforms installed):
             The list of integrations to report the results and logs to. Supported platforms are :obj:`"azure_ml"`,
             :obj:`"comet_ml"`, :obj:`"mlflow"`, :obj:`"tensorboard"` and :obj:`"wandb"`.
+        ddp_find_unused_parameters (:obj:`bool`, `optional`):
+            When using distributed training, the value of the flag :obj:`find_unused_parameters` passed to
+            :obj:`DistributedDataParallel`. Will defaut to :obj:`True` if gradient checkpointing is not used,
+            :obj:`False` otherwise.
```
Comment (stas00) [inline src/transformers/training_args.py:246]:
```suggestion
            :obj:`DistributedDataParallel`. Will default to :obj:`False` if model is ``PreTrainedModel``and  gradient checkpointing is used,
            :obj:`True` otherwise.
```

**Review by stas00 — APPROVED** [review]
Excellent! Thank you!

I think the logic is wrong in both places (once in code and once more in doc). Please validate that my suggestions are correct.

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -425,6 +429,13 @@ class TrainingArguments:
     report_to: Optional[List[str]] = field(
         default=None, metadata={"help": "The list of integrations to report the results and logs to."}
     )
+    ddp_find_unused_parameters: Optional[bool] = field(
+        default=None,
+        metadata={
+            "help": "When using distributed training, the value of the flag `find_unused_parameters` passed to "
```
Comment (stas00) [inline src/transformers/training_args.py:435]: this one is single backquotes, right? not double?

Comment (sgugger) [reply src/transformers/training_args.py:435]: This is for the CLI help, not the sphinx docs :-)

File: src/transformers/training_args.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/training_args.py:246]: Not sure about this one, this just adds the `PreTrainedModel` in the equation but says the same thing. I don't think we need to cover the case where the user is dealing with gradient checkpointing on their side (they probably have their custom training loop if that's the case) so we can leave that out, in which case it's just the same as I did without the negation. Maybe it's clearer this way, will push a commit.

File: src/transformers/training_args.py (hunk 3 above)
Comment (stas00) [reply src/transformers/training_args.py:435]: Ah! what a bummer with this constant switch between sphinx and .md/text :( thank you!

File: src/transformers/training_args.py (hunk 2 above)
Comment (stas00) [reply src/transformers/training_args.py:246]: I see what I you mean, but how can this logic be described if it excludes a key part of `PreTrainedModel`?

If it's not a `PreTrainedModel` it is then `True` always, regardless of `gradient_checkpointing`

Comment (sgugger) [reply src/transformers/training_args.py:246]: There is no `gradient_checkpointing` attribute if the model is not a `PreTrainedModel`, it's a transformers thing.

Comment (stas00) [reply src/transformers/training_args.py]: I see what you mean. Your latest incarnation reads well then. Thank you!

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -761,18 +761,20 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
...
                 output_device=self.args.local_rank,
-                find_unused_parameters=(
-                    not getattr(model.config, "gradient_checkpointing", False)
-                    if isinstance(model, PreTrainedModel)
-                    else True
-                ),
+                find_unused_parameters=find_unused_parameters,
             )
-            # find_unused_parameters breaks checkpointing as per
-            # https://github.com/huggingface/transformers/pull/4659#issuecomment-643356021
```
Comment (LysandreJik) [inline src/transformers/trainer.py:775]: Do we need to remove that comment? It was quite helpful to understand why we had it to `False` for gradient checkpointing in the first place.

**Review by LysandreJik — APPROVED** [review]
LGTM!

File: src/transformers/trainer.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/trainer.py:775]: It's been copied when we set `find_unused_parameters` a few lines above ;-)
